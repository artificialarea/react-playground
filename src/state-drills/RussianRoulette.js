import React from 'react';

// src: https://courses.thinkful.com/react-v1/checkpoint/9#assignment
// alt solution: https://github.com/tomatau/state-drills/blob/master/src/state-drills/RouletteGun.js

class RussianRoulette extends React.Component {
  
  constructor(props) {
    super(props)
    this.state = {
      chamber: null,
      spinningTheChamber: false,
    };

    const { bulletInChamber } = this.props
    console.log(`bullet in chamber: ${bulletInChamber}`)

  }

  static defaultProps = {
    bulletInChamber: 8,
  }

  handleTrigger = () => {
    this.setState({ 
      spinningTheChamber: true, 
    })
    
    this.timer = setTimeout(() => {
      const randomChamber = Math.ceil(Math.random() * 8)
      console.log(`spinning stops at chamber: ${randomChamber}`)
      this.setState({
        chamber: randomChamber,
        spinningTheChamber: false,
      })
    }, 2000)
  }

  componentWillUnmount() {
    clearTimeout(this.timer)
  }

  renderDisplay() {

    const { chamber, spinningTheChamber } = this.state
    // ^^^^^^ destructuring object 
    // equivalent to:
    // const chamber = this.state.chamber;
    // const spinningTheChamber = this.state.spinningTheChamber;

    const { bulletInChamber } = this.props
    // console.log(`bullet in chamber:  ${bulletInChamber}`)

    if (spinningTheChamber) {
      return 'spinning the chamber and pulling the trigger! ...'
    } else if (chamber === bulletInChamber) {
      return 'BANG!!!! Yer dead.'
    } else if (chamber === null) {
      return 'Scene from The Deer Hunter'
    } else {
      return 'Phew! You\'re safe!'
    }
  }
  
  render() {
    return (
      <div className="RouletteGun">
        <p className="status">{this.renderDisplay()}</p>
        <button
          onClick={this.handleTrigger}
        >
          Spin Chamber and Pull the Trigger!
        </button>
      </div>
    )
  }
}

export default RussianRoulette;