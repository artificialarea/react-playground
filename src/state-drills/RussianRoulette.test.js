import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import RussianRoulette from './RussianRoulette';

describe('<RussianRoulette />', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<RussianRoulette />, div)
    ReactDOM.unmountComponentAtNode(div)
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<RussianRoulette />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  });

});