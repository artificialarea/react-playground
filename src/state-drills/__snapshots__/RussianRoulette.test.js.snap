// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RussianRoulette /> renders the UI as expected 1`] = `
<div
  className="RouletteGun"
>
  <p
    className="status"
  >
    Scene from The Deer Hunter
  </p>
  <button
    onClick={[Function]}
  >
    Spin Chamber and Pull the Trigger!
  </button>
</div>
`;
